FirebaseCore/Internal/Sources/HeartbeatLogging/RingBuffer.swift

```diff
@@ -38,14 +39,23 @@ struct RingBuffer<Element>: Sequence {
       return nil
     }
 
-    defer {
-      // Increment index, wrapping around to the start if needed.
-      tailIndex += 1
-      tailIndex %= circularQueue.count
+    if !circularQueue.indices.contains(tailIndex) {
+      // We have somehow entered an invalid state (#10025) and will readjust
+      // our `tailIndex` to be in bounds. Resetting to the `startIndex` is an
+      // arbitrary solution and may result in the  premature overwriting of
+      // existing elements.
+      tailIndex = circularQueue.startIndex
```

Comment:
Is it feasible to add an error heartbeat to the log instead?

Reply:
Good idea. It would be feasible. Instead of trying to handle the behavior in this context, I could throw an error here, catch it in the above HeartbeatsBundle context, and reset the buffer (optionally adding a heartbeat containing an error message). Would you like to see what that looks like? WDYT?

Reply:
That would be great. Then we could track the behavior without impacting app behavior or reliability.

Reply:
Okay, I believe I have arrived at the desired behavior. As of now, I didn't add unit tests for this behavior since it would involve either making a private variable _internal_ so I used conditional breakpoints instead to test the flow. LMK if you feel this strongly that I should explicitly unit test the failing code path.